alg/teca_spectral_filter.py

```diff
@@ -0,0 +1,253 @@
+        if not isinstance(arrays, list):
+            arrays = list(arrays)
+
+        # for each array to filter
+        for array_in in self.point_arrays:
+
+            # remove the result array from the request
+            if self.pass_input_arrays:
+                array_out = array_in + '_' + self.filter_type
+                if array_out in arrays:
+                    arrays.remove(array_out)
+
+            # request the array to filter
+            if not array_in in arrays:
+                arrays.append(array_in)
+
+        # update the list of requested arrays
+        req['arrays'] = arrays
+
+        return [req]
+
+    def execute(self, port, data_in, req):
+        """ TECA execute override """
+        exec_start = time.monotonic_ns()
+
+        try:
+            rank = self.get_communicator().Get_rank()
+        except Exception:
+            rank = 0
+
+        # get the device to execute on, assigned by the execution engine
+        dev = -1
+        npmod = numpy
+        if self.have_cuda:
+            dev = req['device_id']
+            if dev >= 0:
+                npmod = cupy
+                cupy.cuda.Device(dev).use()
+
+        # get the input
+        mesh_in = as_teca_cartesian_mesh(data_in[0])
+
+        # allocate the output and shallow copy to preserve data that we don't process
+        mesh_out = teca_cartesian_mesh.New()
+        mesh_out.shallow_copy(mesh_in)
+
+        # get the time units in seconds
+        t_units = mesh_in.get_time_units()
+        seconds_per = self.get_number_of_seconds(t_units)
+
+        # get the number of time steps and time span of this window
+        t_ext = mesh_in.get_temporal_extent()
+        win_size = t_ext[1] - t_ext[0] + 1
+
+        t_bds = mesh_in.get_temporal_bounds()
+        t_win = t_bds[1] - t_bds[0]
+
+        # compute the sampling rate in Hz
+        sample_rate = win_size / ( t_win * seconds_per )
+
+        # generate the filter kernel
+        high_pass = 1 if self.filter_type == 'high_pass' else 0
+
+        H = self.butterworth(npmod, win_size, sample_rate,
+                        self.critical_frequency, self.filter_order,
+                        high_pass)
+
+        # reshape so that numpy broadcasting works
+        H.shape = [win_size, 1, 1, 1]
+
+        # apply the filter in the frequenmcy domain
+        arrays_in = mesh_in.get_point_arrays()
+        arrays_out = mesh_out.get_point_arrays()
+
+        for array_name_in in self.point_arrays:
+
+            # get the input array
+            if dev < 0:
+                array_in = arrays_in[array_name_in].get_cpu_accessible()
+            else:
+                array_in = arrays_in[array_name_in].get_cuda_accessible()
+
+            # reshape the input array
+            nx,ny,nz,nt = mesh_in.get_array_shape(array_name_in)
+            array_in.shape = (nt,nz,ny,nx)
+
+            # apply the filter in the frequency domain
+            array_out = npmod.fft.ifft( npmod.fft.ifftshift(
```

Comment:
I think the application of the butterworth filter is more complicated than just a simple fft-based filter.  It's a type of filter I'm not too familiar with, though, so I can't comment more.  But to get an idea of the complexity, look at the source code for [scipy.signal.butter](https://github.com/scipy/scipy/blob/v1.8.1/scipy/signal/_filter_design.py#L2935-L3046) and [scipy.signal.sosfilt](https://github.com/scipy/scipy/blob/v1.8.1/scipy/signal/_signaltools.py#L4129-L4238), which are the two routines that would be used to apply the butterworth filter using scipy.

Reply:
See also [DSP Guide Chapter 20](http://www.dspguide.com/ch20/2.htm) for a discussion of this type of filter (which differs from a convolutional filter).

Reply:
What I'm doing is described in chapter 18 of the book your pointing to. What you're pointing to here in chap 20 is about chebychev filers and uses a different approach - the z-transform.

Reply:
I think this is worth looking into these methods, although it's going to take more time than I have before my talk. I think I will certainly emphasize that what I'm doing is not ideal and is only for demonstration purposes.

Reply:
> What I'm doing is described in chapter 18
Yes, I understand that.  The point I was making was that I don't think the butterworth filter is traditionally applied via the convolution theorem, and I think there's a good reason for that (but one that I don't have direct experience with).

Reply:
Also, if I read the code correctly, it looks like the FFT is applied to the data directly without any zero-padding and tapering.  The use of the FFT assumes that the data are periodic--which is true in the test cases that you show. But in real-world data it isn't, and direct application of FFT-based transforms without zero-padding will result data from the end of the timeseries bleeding onto data at the beginning of the timeseries (and vice versa).  Tapering of some kind is necessary when zero-padding because otherwise you're effectively multiplying by a top-hat function in the temporal domain which leads to severe ringing in the spectral domain, which messes up the filter characteristics.

I suggest adding a test that isn't exactly periodic.

Reply:
OK. You're right. tried to test that, but now that I think about it all my window sizes were exact periods of the sin waves.

Reply:
do you know if I use the scipy routines if I'll

> @burlen I can't think of any further comments. Nice work on this!

@taobrienlbl Thanks. I'm having some doubts about weather or not this should be used/presented in this form. I could use scipy signal instead but I do not think that will work on the GPU, and I think I'll still have to deal with the windowing issues.I might be able to run with one window.